### PR TITLE
Ensure superclass is only returned for class creation calls

### LINKED ISSUE #964: Getting "`inspect': undefined method `name' for s(:str, "SuccessResult")" in 4.1.0.
Hi.

After the upgrade from 4.0.2 to 4.1.0, "`inspect': undefined method `name' for s(:str, "SuccessResult")"  has occurred. 

Stack trace is following:

```ruby
$HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/smells/subclassed_from_core_class.rb:34:in `inspect': undefined method `name' for s(:str, "SuccessResult"):#<Class:0x0055ee843368c8> (NoMethodError)
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/smells/smell_detector.rb:47:in `run_for'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/smells/smell_repository.rb:47:in `block in examine'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/smells/smell_repository.rb:46:in `each'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/smells/smell_repository.rb:46:in `examine'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/examiner.rb:81:in `block in run'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/context/code_context.rb:85:in `each'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/context/code_context.rb:87:in `block in each'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/context/code_context.rb:86:in `each'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/context/code_context.rb:86:in `each'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/context/code_context.rb:87:in `block in each'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/context/code_context.rb:86:in `each'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/context/code_context.rb:86:in `each'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/examiner.rb:80:in `run'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/reek-4.1.0/lib/reek/examiner.rb:37:in `initialize'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rubycritic-2.9.2/lib/rubycritic/analysers/smells/reek.rb:28:in `new'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rubycritic-2.9.2/lib/rubycritic/analysers/smells/reek.rb:28:in `add_smells_to'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rubycritic-2.9.2/lib/rubycritic/analysers/smells/reek.rb:15:in `block in run'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rubycritic-2.9.2/lib/rubycritic/core/analysed_modules_collection.rb:25:in `each'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rubycritic-2.9.2/lib/rubycritic/core/analysed_modules_collection.rb:25:in `each'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rubycritic-2.9.2/lib/rubycritic/analysers/smells/reek.rb:14:in `run'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rubycritic-2.9.2/lib/rubycritic/analysers_runner.rb:28:in `block in run'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rubycritic-2.9.2/lib/rubycritic/analysers_runner.rb:25:in `each'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rubycritic-2.9.2/lib/rubycritic/analysers_runner.rb:25:in `run'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rubycritic-2.9.2/lib/rubycritic/commands/default.rb:22:in `critique'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rubycritic-2.9.2/lib/rubycritic/commands/default.rb:17:in `execute'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rubycritic-2.9.2/lib/rubycritic/cli/application.rb:18:in `execute'
	from $HOME/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rubycritic-2.9.2/bin/rubycritic:9:in `<top (required)>'
	from $HOME/.rbenv/versions/2.2.3/bin/rubycritic:23:in `load'
	from $HOME/.rbenv/versions/2.2.3/bin/rubycritic:23:in `<main>'
```  

Do I need something for upgrade?  Thanks.

### PR DESCRIPTION
**Pull request by mvz**
Fixes #964.

### CODE REVIEW

**Comment by mvz** [comment]
Anyone up for reviewing this?

**Comment by troessner** [comment]
I can take a look tonight!

File: lib/reek/ast/sexp_extensions/module.rb [hunk 1]
```diff
@@ -58,6 +58,47 @@ def defines_module?
...
+        #
+        # source = "class Hi; THIS ||= 3; end"
+        # (class
+        #   (const nil :Hi) nil
+        #   (or-asgn
+        #    (casgn nil :THIS)
+        #    (int 3)))
+        def value
+          children[2]
+        end
```
Comment (troessner) [inline lib/reek/ast/sexp_extensions/module.rb]: The diff is confusing here. Why are the `value` and the `name` method in here? They didn't change. So the only change is that now they are public? Why?

File: spec/reek/ast/sexp_extensions_spec.rb [hunk 2]
```diff
@@ -441,7 +441,7 @@
 end
 
 RSpec.describe Reek::AST::SexpExtensions::CasgnNode do
-  context '#defines_module' do
+  context '#defines_module?' do
```
Comment (troessner) [inline spec/reek/ast/sexp_extensions_spec.rb]: Since we're at it, shouldn't this be a "describe"?

File: spec/reek/ast/sexp_extensions_spec.rb [hunk 3]
```diff
@@ -474,6 +474,14 @@
         expect(exp.defines_module?).to be_falsey
       end
     end
+
+    context 'when assign a string to a constant' do
```
Comment (troessner) [inline spec/reek/ast/sexp_extensions_spec.rb]: s/assign/assigning ;)

File: spec/reek/ast/sexp_extensions_spec.rb [hunk 4]
```diff
@@ -494,33 +502,17 @@
 
       expect(exp.superclass).to be_nil
     end
-  end
 
-  context '#constant_definition' do
-    context 'simple constant assignment' do
-      it 'is not important' do
```
Comment (troessner) [inline spec/reek/ast/sexp_extensions_spec.rb]: Now that is a great example description 😆

File: spec/reek/smells/subclassed_from_core_class_spec.rb [hunk 5]
```diff
@@ -130,6 +130,28 @@ class Dummy
...
+
+  it 'should only report classes created with Class.new' do
+    src = <<-EOS
+      module Namespace
+        class Dummy
+          Dummiest = Foo.new(Array)
+        end
+      end
+    EOS
+    expect(src).to_not reek_of(:SubclassedFromCoreClass)
```
Comment (troessner) [inline spec/reek/smells/subclassed_from_core_class_spec.rb]: Very readable specs, like it!

File: spec/reek/ast/sexp_extensions_spec.rb (hunk 2 above)
Comment (mvz) [inline spec/reek/ast/sexp_extensions_spec.rb]: Yes, I missed that :smile:.

File: lib/reek/ast/sexp_extensions/module.rb (hunk 1 above)
Comment (mvz) [inline lib/reek/ast/sexp_extensions/module.rb]: Oh, the diff is very confusing: Name and value didn't change. It's the methods below that became private.

Comment (troessner) [inline lib/reek/ast/sexp_extensions/module.rb]: Ah, ok!

**Comment by troessner** [comment]
A bunch of typo fixes and this honeybadger is ready to roll!

**Comment by mvz** [comment]
Updated, @troessner.

**Pull request merged**